more docs (#57)
equinor/template-fastapi-react: docs/how-to/using-git.md

---
title: Using Git
---

Using Git
=============

Git is used as version control system for the source code in the project.

## Basic principles

> Whenever the devleoper makes any code changes, a new Pull Request (PR) should be created. The new PR should branch out from the main branch.

Reviewer: ```suggestion
Whenever a developer makes changes to the code, a new Pull Request (PR) should be created. The new PR should branch out from the main branch.
```

> Whenever the devleoper makes any code changes, a new Pull Request (PR) should be created. The new PR should branch out from the main branch.

Reviewer 2: developer

Reviewer 2: Also, not sure if I agree that a new PR should always branch out from main. It can be very useful to branch out separate issues from a topic branch and create a PR into the given branch instead of just making a single (huge) PR that contains multiple features/fixes.

Ideally, each PR will do one thing (one new feature, fix one bug, etc). Keeping PRs small will simplify the code review process significantly. Also, a PR should be connected to an active issue from the project board.

Tip: on Github, if you write the syntax "closes #XXX" (XXX is the issue number) in the PR description, the connected issue will be automatically closed when the PR is merged.

> When a branch (feature) is done, the developers create a PR and asks for review. When a PR is approved, the branch is merged to main. At least one person should approve the PR before merging to main. Merging will trigger the [github actions workflow](github-actions.md) which automatically tests and deploys.

Reviewer 2: `When a PR is approved` What if the PR isn't approved? Or do we take it for granted? Ideally we should have some guidelines for when to close a PR, if it's growing too complex or has gotten outdated due to staying open for too long. A PR can affect the effectiveness of the team badly if it remains open.

## Branch naming
The branch should follow the naming structure:

```
<type>/<what-my-pr-does>
```
The type can be one of these types:
* feat: a feature
* fix: a bugfix
* build: a change to build system or external dependencies
* ci: a change to continous integration/delivery setup (typically scripts and config - files)
* docs: a changes to documentation
* style: a change to style 
* refactor: a code change that neither fixes a bug nor adds/removes a feature
* test: adding tests, refactoring test; no production code change
* chore: updating build tasks, package manager configs, etc; no production code change

> * perf: a code change that improves performance

Reviewer 2: Do we need all these types? Only `fix` and `feat` are mandated by the conventional commits specification. I've only found half of these types useful. If the type isn't a fix or feat I always find the prefix the least useful part of the commit message/branch name. Am I the only one?

## Commits

> Keep commit messages informative.

Reviewer 2: Maybe we can add some examples or do's/dont's? For instance messages like: `fixed comments from review` may seem informative within the scope of the PR, but are not really useful when you need to debug or go through the git history later.

Here is a template for commit messages:

```

> <type>(<optional scope>): imperative tense summary, <= 50 chars

Reviewer: What does <optional scope> refer to in this context?

Reviewer: Also, it is not evident if type refers to the same types as above, or if this type parameter includes more or less types than above.

> When necessarry, mote details can come here, until 72 chars each line.

Reviewer: typo: necessary. Rephrase: "When necessary" -> "If necessary"; "until 72" -> "up to 72 on each line"

Refs/Resolves #<issue id>

BREAKING CHANGE: <description>
```

Also, remember to set up [Pre-commit](pre-commit.md).

> ## What to check before you send a PR to code review

Reviewer 2: This section and `What to check in code reviews` should be merged. The team share equal responsibility for the code, and we should review our own code before asking team members.

> * You use conventional commit messages

Reviewer: The commit messages above are not described as "conventional" meaning that the word conventional in this list loses meaning unless you already know what it means. I suggest adding the word conventional in the paragraph above: "Commit messages should follow the conventional commit message specification". We could also consider linking to the documentation of conventional commit there-

> * Unit tests passes

Reviewer 3: ```suggestion
* All tests passes
```

> * Other automatic tests passes, given it is feasible to execute them before code review 

Reviewer 3: ```suggestion
```

* Naming conventions

> ## What to check in code reviews

Reviewer 2: Maybe we can list the SOLID principles, to have something thats more easily measured than `is the code well-designed?`

* Functionality: Does the code behave as the PR author likely intended? Does the code behave as users would expect? If relevant, test the new feature manually on your local machine and try edge cases that might break the code.
* Software design: Is the code well-designed and fitted to the surrounding architecture?

> * Complexity: Would another developer be able to easily understand and use the code?

Reviewer 3: ```suggestion
* Readability: Would another developer be able to easily understand and work with the code? Are there none-idiomatic patterns?
```

* Tests: Does the PR have correct and well-designed automated tests?

> * Naming: Are names for variables, functions, etc. descriptive?

Reviewer 2: + following correct naming conventions, casing, aim for names to be _exact_ - makes it easier to avoid violation of the the single responsibility principle.

> * Comments: Are the comments clear and useful?

Reviewer: ```suggestion
* Comments: Are comments understandable and useful?
```
